File: src/analytics.py
"""
本地统计分析 - 基于已有 CSV 观看记录生成用户画像
无需额外网络请求，仅使用标准库。
"""
from collections import Counter, defaultdict
from datetime import datetime, timedelta
from typing import List, Dict, Any


def _parse_view_time(item: dict) -> datetime:
    """解析观看时间字符串"""
    try:
        return datetime.strptime(item.get("观看时间", ""), "%Y-%m-%d %H:%M:%S")
    except Exception:
        return datetime.min


def _author_name(item: dict) -> str:
    """统一获取作者/主播/UP 主名称"""
    return (item.get("UP主") or item.get("主播") or item.get("作者") or "未知作者").strip()


def _content_type_name(item: dict) -> str:
    """内容类型中文显示"""
    mapping = {"video": "视频", "live": "直播", "article": "专栏"}
    return mapping.get(item.get("类型", ""), item.get("类型", "其他"))


def _duration_seconds(item: dict) -> int:
    """获取总时长（秒），仅视频有意义"""
    try:
        return int(item.get("总时长(秒)", 0) or 0)
    except (TypeError, ValueError):
        return 0


def _progress_seconds(item: dict) -> int:
    """获取已观看时长（秒）"""
    try:
        progress = int(item.get("已观看(秒)", 0) or 0)
        # -1 在游标接口表示已看完，这里已经过 parser 转换，兼容兜底
        if progress < 0:
            return _duration_seconds(item)
        return progress
    except (TypeError, ValueError):
        return 0


def _completion_ratio(item: dict) -> float:
    """计算单条完成度 0.0~1.0"""
    duration = _duration_seconds(item)
    progress = _progress_seconds(item)
    if duration <= 0:
        return 0.0
    return min(1.0, progress / duration)


def compute_basic_stats(records: List[dict]) -> Dict[str, Any]:
    """计算基础统计数字"""
    total = len(records)
    if total == 0:
        return {
            "total_records": 0,
            "total_videos": 0,
            "total_lives": 0,
            "total_articles": 0,
            "total_watch_seconds": 0,
            "total_watch_time_text": "0分钟",
            "unique_authors": 0,
            "avg_completion": "0%",
        }

    type_counter = Counter(_content_type_name(r) for r in records)
    watch_seconds = sum(_progress_seconds(r) for r in records)

    # 平均完播率：仅统计有总时长的视频
    video_ratios = [_completion_ratio(r) for r in records if _duration_seconds(r) > 0]
    avg_completion = (sum(video_ratios) / len(video_ratios) * 100) if video_ratios else 0

    return {
        "total_records": total,
        "total_videos": type_counter.get("视频", 0),
        "total_lives": type_counter.get("直播", 0),
        "total_articles": type_counter.get("专栏", 0),
        "total_watch_seconds": watch_seconds,
        "total_watch_time_text": _format_duration(watch_seconds),
        "unique_authors": len(set(_author_name(r) for r in records)),
        "avg_completion": f"{avg_completion:.1f}%",
    }


def top_authors(records: List[dict], top_n: int = 10) -> List[Dict[str, Any]]:
    """最常看 UP 主：次数 + 累计观看时长 + 平均完播率"""
    author_records = defaultdict(list)
    for r in records:
        author = _author_name(r)
        if author and author != "未知作者":
            author_records[author].append(r)

    result = []
    for author, items in author_records.items():
        watch_seconds = sum(_progress_seconds(r) for r in items)
        ratios = [_completion_ratio(r) for r in items if _duration_seconds(r) > 0]
        avg_completion = (sum(ratios) / len(ratios)) if ratios else 0
        result.append({
            "name": author,
            "count": len(items),
            "watch_seconds": watch_seconds,
            "watch_time_text": _format_duration(watch_seconds),
            "avg_completion": avg_completion,
        })

    # 按观看次数降序，次数相同按时长降序
    result.sort(key=lambda x: (x["count"], x["watch_seconds"]), reverse=True)
    return result[:top_n]


def top_categories(records: List[dict], top_n: int = 10) -> List[Dict[str, Any]]:
    """最常看分类：次数 + 累计时长"""
    category_records = defaultdict(list)
    for r in records:
        category = (r.get("分类") or "未分类").strip()
        category_records[category].append(r)

    result = []
    for category, items in category_records.items():
        watch_seconds = sum(_progress_seconds(r) for r in items)
        result.append({
            "name": category,
            "count": len(items),
            "watch_seconds": watch_seconds,
            "watch_time_text": _format_duration(watch_seconds),
        })

    result.sort(key=lambda x: (x["count"], x["watch_seconds"]), reverse=True)
    return result[:top_n]


def time_of_day_distribution(records: List[dict]) -> Dict[str, int]:
    """观看时段分布：凌晨/上午/下午/晚上"""
    buckets = {"凌晨": 0, "上午": 0, "下午": 0, "晚上": 0}
    for r in records:
        dt = _parse_view_time(r)
        if dt == datetime.min:
            continue
        hour = dt.hour
        if 0 <= hour < 6:
            buckets["凌晨"] += 1
        elif 6 <= hour < 12:
            buckets["上午"] += 1
        elif 12 <= hour < 18:
            buckets["下午"] += 1
        else:
            buckets["晚上"] += 1
    return buckets


def top_authors_by_completion(records: List[dict], min_count: int = 3, top_n: int = 10) -> List[Dict[str, Any]]:
    """完播率最高的 UP 主（至少看过 min_count 次，避免只看一次就 100% 的偏差）"""
    author_records = defaultdict(list)
    for r in records:
        author = _author_name(r)
        if author and author != "未知作者" and _duration_seconds(r) > 0:
            author_records[author].append(r)

    result = []
    for author, items in author_records.items():
        if len(items) < min_count:
            continue
        ratios = [_completion_ratio(r) for r in items]
        avg_completion = sum(ratios) / len(ratios)
        result.append({
            "name": author,
            "count": len(items),
            "avg_completion": avg_completion,
            "avg_completion_text": f"{avg_completion * 100:.1f}%",
        })

    result.sort(key=lambda x: x["avg_completion"], reverse=True)
    return result[:top_n]


def daily_trend(records: List[dict], days: int = 30) -> List[Dict[str, Any]]:
    """最近 N 天每日观看数量趋势"""
    if not records:
        return []

    today = datetime.now().replace(hour=0, minute=0, second=0, microsecond=0)
    counts = {today - timedelta(days=i): 0 for i in range(days)}

    for r in records:
        dt = _parse_view_time(r)
        if dt == datetime.min:
            continue
        day = dt.replace(hour=0, minute=0, second=0, microsecond=0)
        if day in counts:
            counts[day] += 1

    return [
        {"date": day.strftime("%m-%d"), "count": count}
        for day, count in sorted(counts.items())
    ]


def _format_duration(total_seconds: int) -> str:
    """把秒数格式化为易读文本"""
    if total_seconds < 60:
        return f"{total_seconds}秒"
    if total_seconds < 3600:
        return f"{total_seconds // 60}分钟"
    if total_seconds < 86400:
        hours = total_seconds / 3600
        return f"{hours:.1f}小时"
    days = total_seconds / 86400
    return f"{days:.1f}天"

File: src/business/analytics.cpp
#include "analytics.h"

#include <QDate>
#include <QHash>
#include <QSet>

#include <algorithm>
#include <functional>
#include <vector>

namespace bili::business {

namespace {

struct Aggregate {
    int count = 0;
    qint64 watchSeconds = 0;
    qint64 completionSum = 0;
};

qint64 watchedSecondsForRecord(const RecordPtr& record)
{
    return record ? record->watchedSeconds() : 0;
}

int completionForRecord(const RecordPtr& record)
{
    return record ? record->effectiveCompletionPercent() : 0;
}

QVariantMap aggregateToMap(const QString& name, const Aggregate& agg)
{
    QVariantMap map;
    map[QStringLiteral("name")] = name;
    map[QStringLiteral("count")] = agg.count;
    map[QStringLiteral("watch_seconds")] = static_cast<qlonglong>(agg.watchSeconds);
    map[QStringLiteral("watch_time_text")] = formatDuration(static_cast<int>(agg.watchSeconds));
    const int avgCompletion = agg.count > 0 ? static_cast<int>(agg.completionSum / agg.count) : 0;
    map[QStringLiteral("avg_completion")] = avgCompletion;
    map[QStringLiteral("avg_completion_text")] = QStringLiteral("%1%").arg(avgCompletion);
    return map;
}

template <typename KeyFunc>
QHash<QString, Aggregate> aggregateBy(const RecordList& records, KeyFunc keyFunc)
{
    QHash<QString, Aggregate> map;
    map.reserve(static_cast<int>(records.size()));
    for (const auto& record : records) {
        if (!record) continue;
        const QString key = keyFunc(record.get());
        if (key.isEmpty()) continue;
        Aggregate& agg = map[key];
        ++agg.count;
        agg.watchSeconds += watchedSecondsForRecord(record);
        agg.completionSum += completionForRecord(record);
    }
    return map;
}

using RankedItem = std::pair<QString, Aggregate>;

// 默认过滤谓词：接受所有条目。仅作为 topNFromMap 的模板默认参数使用。
struct AcceptAllAggregate {
    bool operator()(const Aggregate&) const noexcept { return true; }
};

template <typename Cmp, typename Pred = AcceptAllAggregate>
QVariantList topNFromMap(QHash<QString, Aggregate> map,
                         int topN,
                         Cmp cmp,
                         Pred pred = Pred{})
{
    std::vector<RankedItem> list;
    list.reserve(static_cast<size_t>(map.size()));
    for (auto it = map.begin(); it != map.end(); ++it) {
        if (pred(it.value())) {
            list.emplace_back(it.key(), it.value());
        }
    }

    const auto less = [&cmp](const RankedItem& a, const RankedItem& b) {
        return cmp(a.second, b.second); // true if a should come before b
    };

    if (topN > 0 && static_cast<size_t>(topN) < list.size()) {
        std::partial_sort(list.begin(), list.begin() + topN, list.end(), less);
        list.resize(static_cast<size_t>(topN));
    } else {
        std::sort(list.begin(), list.end(), less);
    }

    QVariantList result;
    result.reserve(static_cast<int>(list.size()));
    for (const auto& item : list) {
        result.append(aggregateToMap(item.first, item.second));
    }
    return result;
}

} // namespace

QString formatDuration(int totalSeconds)
{
    if (totalSeconds <= 0) {
        return QStringLiteral("0秒");
    }

    const int seconds = totalSeconds % 60;
    const int minutes = (totalSeconds / 60) % 60;
    const int hours = (totalSeconds / 3600) % 24;
    const int days = totalSeconds / 86400;

    if (days > 0) {
        return QStringLiteral("%1天%2小时%3分%4秒")
            .arg(days)
            .arg(hours)
            .arg(minutes)
            .arg(seconds);
    }
    if (hours > 0) {
        return QStringLiteral("%1小时%2分%3秒").arg(hours).arg(minutes).arg(seconds);
    }
    if (minutes > 0) {
        return QStringLiteral("%1分%2秒").arg(minutes).arg(seconds);
    }
    return QStringLiteral("%1秒").arg(seconds);
}

QVariantMap computeBasicStats(const RecordList& records)
{
    int totalVideos = 0;
    int totalLives = 0;
    int totalArticles = 0;
    qint64 totalWatchSeconds = 0;
    qint64 completionSum = 0;
    QSet<QString> authors;

    for (const auto& record : records) {
        if (!record) continue;

        switch (record->type) {
        case RecordType::Video:   ++totalVideos; break;
        case RecordType::Live:    ++totalLives; break;
        case RecordType::Article: ++totalArticles; break;
        default: break;
        }

        totalWatchSeconds += watchedSecondsForRecord(record);
        completionSum += completionForRecord(record);
        authors.insert(record->authorName);
    }

    QVariantMap stats;
    stats[QStringLiteral("total_records")] = static_cast<int>(records.size());
    stats[QStringLiteral("total_videos")] = totalVideos;
    stats[QStringLiteral("total_lives")] = totalLives;
    stats[QStringLiteral("total_articles")] = totalArticles;
    stats[QStringLiteral("total_watch_seconds")] = static_cast<qlonglong>(totalWatchSeconds);
    stats[QStringLiteral("total_watch_time_text")] = formatDuration(static_cast<int>(totalWatchSeconds));
    stats[QStringLiteral("unique_authors")] = static_cast<int>(authors.size());
    const int avgCompletion = records.empty() ? 0 : static_cast<int>(completionSum / static_cast<int>(records.size()));
    stats[QStringLiteral("avg_completion")] = avgCompletion;
    stats[QStringLiteral("avg_completion_text")] = QStringLiteral("%1%").arg(avgCompletion);
    return stats;
}

QVariantList topAuthors(const RecordList& records, int topN)
{
    return topNFromMap(
        aggregateBy(records, [](const BaseRecord* r) { return r->authorName; }),
        topN,
        [](const Aggregate& a, const Aggregate& b) {
            if (a.count != b.count) return a.count > b.count;
            return a.watchSeconds > b.watchSeconds;
        });
}

QVariantList topCategories(const RecordList& records, int topN)
{
    return topNFromMap(
        aggregateBy(records, [](const BaseRecord* r) { return r->category; }),
        topN,
        [](const Aggregate& a, const Aggregate& b) {
            if (a.count != b.count) return a.count > b.count;
            return a.watchSeconds > b.watchSeconds;
        });
}

QVariantMap timeOfDayDistribution(const RecordList& records)
{
    int dawn = 0;      // 00:00 - 05:59
    int morning = 0;   // 06:00 - 11:59
    int afternoon = 0; // 12:00 - 17:59
    int evening = 0;   // 18:00 - 23:59

    for (const auto& record : records) {
        if (!record || !record->viewAt.isValid()) continue;
        const int hour = record->viewAt.time().hour();
        if (hour < 6) {
            ++dawn;
        } else if (hour < 12) {
            ++morning;
        } else if (hour < 18) {
            ++afternoon;
        } else {
            ++evening;
        }
    }

    QVariantMap map;
    map[QStringLiteral("凌晨")] = dawn;
    map[QStringLiteral("上午")] = morning;
    map[QStringLiteral("下午")] = afternoon;
    map[QStringLiteral("晚上")] = evening;
    return map;
}

QVariantList topAuthorsByCompletion(const RecordList& records, int minCount, int topN)
{
    auto map = aggregateBy(records, [](const BaseRecord* r) { return r->authorName; });

    const auto cmp = [](const Aggregate& a, const Aggregate& b) {
        const int avgA = a.count > 0 ? static_cast<int>(a.completionSum / a.count) : 0;
        const int avgB = b.count > 0 ? static_cast<int>(b.completionSum / b.count) : 0;
        if (avgA != avgB) return avgA > avgB;
        return a.count > b.count;
    };

    const auto pred = [minCount](const Aggregate& a) {
        return a.count >= minCount;
    };

    return topNFromMap(std::move(map), topN, cmp, pred);
}

QVariantList dailyTrend(const RecordList& records, int days)
{
    const QDate today = QDate::currentDate();
    const QDate windowStart = today.addDays(-(days - 1));
    QHash<QDate, int> counts;
    counts.reserve(static_cast<int>(records.size()));

    for (const auto& record : records) {
        if (!record || !record->viewAt.isValid()) continue;
        const QDate date = record->viewAt.date();
        // 仅聚合 [windowStart, ∞) 区间：未来日期（时钟漂移/数据预录）也计入，
        // 不再用 date.daysTo(today) < days，避免未来记录被错误排除
        if (date.isValid() && date >= windowStart) {
            ++counts[date];
        }
    }

    QVariantList result;
    result.reserve(days);
    for (int i = days - 1; i >= 0; --i) {
        const QDate date = today.addDays(-i);
        QVariantMap map;
        map[QStringLiteral("date")] = date.toString(QStringLiteral("yyyy-MM-dd"));
        map[QStringLiteral("count")] = counts.value(date, 0);
        result.append(map);
    }
    return result;
}

QVariantList monthlyTrend(const RecordList& records, int months)
{
    const QDate today = QDate::currentDate();
    QHash<QString, int> counts;
    counts.reserve(static_cast<int>(records.size()));

    for (const auto& record : records) {
        if (!record || !record->viewAt.isValid()) continue;
        const QDate date = record->viewAt.date();
        if (date.isValid()) {
            const QString key = date.toString(QStringLiteral("yyyy-MM"));
            ++counts[key];
        }
    }

    QVariantList result;
    result.reserve(months);
    for (int i = months - 1; i >= 0; --i) {
        const QDate month = today.addMonths(-i);
        const QString key = month.toString(QStringLiteral("yyyy-MM"));
        QVariantMap map;
        map[QStringLiteral("date")] = key;
        map[QStringLiteral("count")] = counts.value(key, 0);
        result.append(map);
    }
    return result;
}

QVariantList yearlyTrend(const RecordList& records)
{
    QHash<int, int> counts;
    counts.reserve(static_cast<int>(records.size()));

    for (const auto& record : records) {
        if (!record || !record->viewAt.isValid()) continue;
        const QDate date = record->viewAt.date();
        if (date.isValid()) {
            ++counts[date.year()];
        }
    }

    const QList<int> years = counts.keys();
    const int minYear = years.isEmpty() ? QDate::currentDate().year() : *std::min_element(years.begin(), years.end());
    const int maxYear = years.isEmpty() ? QDate::currentDate().year() : *std::max_element(years.begin(), years.end());

    QVariantList result;
    for (int year = minYear; year <= maxYear; ++year) {
        QVariantMap map;
        map[QStringLiteral("date")] = QString::number(year);
        map[QStringLiteral("count")] = counts.value(year, 0);
        result.append(map);
    }
    return result;
}

} // namespace bili::business
